source/ConditionsStore.cpp

```diff
@@ -0,0 +1,456 @@
+}
+
+
+
+// Constructor where a number of initial manually-set values are set.
+ConditionsStore::ConditionsStore(initializer_list<pair<string, int64_t>> initialConditions)
+{
+	for(const auto &it : initialConditions)
+		Set(it.first, it.second);
+}
+
+
+
+// Constructor where a number of initial manually-set values are set.
+ConditionsStore::ConditionsStore(const map<string, int64_t> &initialConditions)
+{
+	for(const auto &it : initialConditions)
+		Set(it.first, it.second);
+}
+
+
+
+void ConditionsStore::Load(const DataNode &node)
+{
+	for(const DataNode &child : node)
+		Set(child.Token(0), (child.Size() >= 2) ? child.Value(1) : 1);
+}
+
+
+
+void ConditionsStore::Save(DataWriter &out) const
+{
+	if(PrimariesBegin() != PrimariesEnd())
+	{
+		out.Write("conditions");
+		out.BeginChild();
+		{
+			for(auto it = PrimariesBegin(); it != PrimariesEnd(); ++it)
+			{
+				// If the condition's value is 1, don't bother writing the 1.
+				if(it->second == 1)
+					out.Write(it->first);
+				else if(it->second)
+					out.Write(it->first, it->second);
+			}
+		}
+		out.EndChild();
+	}
+}
+
+
+
+// Get a condition from the Conditions-Store. Retrieves both conditions
+// that were directly set (primary conditions) as well as conditions
+// derived from other data-structures (derived conditions).
+int64_t ConditionsStore::Get(const string &name) const
+{
+	const ConditionEntry *ce = GetEntry(name);
+	if(!ce)
+		return 0;
+
+	if(!ce->provider)
+		return ce->value;
+
+	return ce->provider->getFun(name);
+}
+
+
+
+bool ConditionsStore::Has(const string &name) const
+{
+	const ConditionEntry *ce = GetEntry(name);
+	if(!ce)
+		return false;
+
+	if(!ce->provider)
+		return true;
+
+	return ce->provider->hasFun(name);
+}
+
+
+
+bool ConditionsStore::empty() const
+{
+	return storage.empty();
+}
+
+
+
+// Add a value to a condition. Returns true on success, false on failure.
+bool ConditionsStore::Add(const string &name, int64_t value)
+{
+	// This code performers 2 lookups of the condition, once for get and
+	// once for set. This might be optimized to a single lookup in a
+	// later version of the code.
+	return Set(name, Get(name) + value);
+}
+
+
+
+// Set a value for a condition, either for the local value, or by performing
+// a set on the provider.
+bool ConditionsStore::Set(const string &name, int64_t value)
+{
+	ConditionEntry *ce = GetEntry(name);
+	if(!ce)
+	{
+		(storage[name]).value = value;
+		return true;
+	}
+	if(!ce->provider)
+	{
+		ce->value = value;
+		return true;
+	}
+	return ce->provider->setFun(name, value);
+}
+
+
+
+// Erase a condition completely, either the local value or by performing
+// an erase on the provider.
+bool ConditionsStore::Erase(const string &name)
+{
+	ConditionEntry *ce = GetEntry(name);
+	if(!ce)
+		return true;
+
+	if(!(ce->provider))
+	{
+		storage.erase(name);
+		return true;
+	}
+	return ce->provider->eraseFun(name);
+}
+
+
+
+ConditionsStore::ConditionEntry &ConditionsStore::operator[](const std::string &name)
```

Comment:
Style change: the file is already using namespace std. There are various other instances of `std::` in this file, but those are on standard function calls, in which case we typically do prepend `std::` to those.
```suggestion
ConditionsStore::ConditionEntry &ConditionsStore::operator[](const string &name)
```